crates/core/src/cluster_state.rs

```diff
@@ -0,0 +1,463 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use ahash::HashMap;
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use tokio::sync::watch;
+
+use restate_types::{GenerationalNodeId, NodeId, PlainNodeId};
+
+type Generation = u32;
+
+#[derive(Debug, Default, Clone, Copy, Eq, PartialEq, Hash)]
+#[repr(u8)]
+pub enum NodeState {
+    #[default]
+    Dead = 0,
+    Alive,
+    FailingOver,
+}
+
+impl NodeState {
+    pub fn is_alive(self) -> bool {
+        matches!(self, Self::Alive | Self::FailingOver)
+    }
+}
+
+#[derive(Debug, Clone, Copy, Default)]
+struct State {
+    generation: Generation,
+    state: NodeState,
+}
+
+static_assertions::assert_eq_size!(State, u64);
+
+#[derive(Debug, Default)]
+struct Node {
+    state: watch::Sender<State>,
+}
+
+#[derive(Default)]
+struct Inner {
+    nodes: RwLock<HashMap<PlainNodeId, Node>>,
```

Comment:
Is parkinglot::RwLock<HashMap> performing better than `DashMap`?

Reply:
Subject to change in the future, but the ability to lock the entire map is currently needed.